cli/src/diff_util.rs

```diff
@@ -387,13 +387,20 @@ fn show_color_words_diff_line(
                 formatter.write_all(data)?;
             }
             DiffHunk::Different(data) => {
-                let before = data[0];
-                let after = data[1];
-                if !before.is_empty() {
-                    formatter.with_label("removed", |formatter| formatter.write_all(before))?;
-                }
-                if !after.is_empty() {
-                    formatter.with_label("added", |formatter| formatter.write_all(after))?;
+                let labels = ["removed", "added"];
+                for (index, label) in labels.iter().enumerate() {
+                    let hunk_part = data[index];
+                    if !hunk_part.is_empty() {
+                        formatter.with_label(label, |formatter| {
+                            if hunk_part.iter().all(u8::is_ascii_whitespace) {
```

Comment:
nit: maybe it's better to adjust `default_refinement()` to refine "nonword" and "whitespace" characters separately, but the diff output looks pretty good without the change. I think this can be addressed separately.

@martinvonz ?

Reply:
Do you have an example where that would help?

Reply:
Indentation changes in a6d470db rendered better, but that might be because contiguous whitespaces are weighted more than split non-word symbol chars. I don't fully understand the refinement function, so it might be wrong.

Reply:
The refinement works on regions of the file that have non-matching regions after previous rounds of refinement. It takes tokens (e.g. words) found in those regions and tries to match them up. What remains between those matching regions remain for the next refinement step [^1]. If we do a non-word-non-whitespace refinement before a whitespace refinement, then we would better line up the `#` in `#<SP>#<TAB>#` vs `#<TAB>#<SP>#` (for better or worse) compared to the current behavior which is to first line up spaces and tabs (because histogram diff starts by lining up less frequent tokens).

Maybe that explains the difference you're seeing. To be clear, I'm not against it. I'm just wondering if it actually improves diffs in general.

[^1]: But the histogram diff algorithm itself also recursive descends into non-matching regions. That's done with the same tokenizer before the refinement step.